### PR TITLE
BB-2240: Custom page editor

### PR DESCRIPTION
**Pull request by giovannicimolin**
This PR adds the Custom Page editor functionality in the new Ocim UX.

**Testing Instructions:**
1. Check out this branch.
2. Rebuild API client.
3. Run Ocim and new frontend.
4. Go to the custom pages and test around.

Some edge cases for testing:
* Change the page content, them click on a link for another page. 
* Go to some page that has some text (http://localhost:3000/console/custom-pages/honor), do a browser refresh and then try to save the content without changing.
* Do the same as above, but change the content.
* Try erasing values.

**Reviewer:**
- [ ] @Agrendalath

### CODE REVIEW

File: frontend/src/console/components/CustomizationSideMenu/CustomizationSideMenu.tsx [hunk 1]
```diff
@@ -12,12 +12,30 @@ export const CustomizationSideMenu: React.FC = () => {
...
-  if (currentLocation.includes('custom_pages')) {
+  if (currentLocation.includes('custom-pages')) {
     activeKey = 1;
   } else if (currentLocation.includes('settings')) {
     activeKey = 2;
   }
 
+  function capitalizeFirstLetter(str: string) {
+    return str[0].toUpperCase() + str.slice(1);
+  }
```
[the remark this answers is not in the document]
Comment (giovannicimolin) [reply frontend/src/console/components/CustomizationSideMenu/CustomizationSideMenu.tsx:23]: Sure!

File: frontend/src/console/components/CustomPages/CustomPages.tsx [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+              <Col className="page-switch-column">
+                <Form className="page-switch-form">
+                  <Form.Label htmlFor="show-page-switch">
+                    <WrappedMessage messages={messages} id="showPage" />
+                  </Form.Label>
+                  <Form.Check
+                    disabled
+                    checked
+                    type="switch"
+                    label=""
```
[the remark this answers is not in the document]
Comment (giovannicimolin) [reply frontend/src/console/components/CustomPages/CustomPages.tsx:77]: The label is hardcoded in the right side of the switch, and the mockups show it on the left side. 
It also looks better that way, and I didn't find an easy way of making the label switch sides with the switch in CSS.

File: frontend/package.json [hunk 3]
```diff
@@ -76,8 +76,10 @@
   "dependencies": {
     "@fortawesome/fontawesome-free": "^5.12.0",
     "@types/react-color": "^3.0.1",
+    "@types/react-rte": "^0.16.1",
     "history": "^4.10.1",
     "immutability-helper": "^3.0.1",
-    "react-color": "^2.18.0"
+    "react-color": "^2.18.0",
+    "react-rte": "^0.16.1"
```
[the remark this answers is not in the document]
Comment (giovannicimolin) [reply frontend/package.json:83]: Nice catch! I'll switch to this editor (initially I looked at Quill.js, but it didn't look as good).
I'll switch to this other editor, I as already found I few issues with `react-rte` aside from what you mentioned.

Btw, we're not supporting custom image upload, because it would be too hard to store and them pass this to the appserver. So I'll only implement the capability to embed externally hosted images.

[the remark this answers is not in the document]
Comment (giovannicimolin) [reply frontend/package.json:83]: @Agrendalath I've found 2 good options:

1. https://github.com/ianstormtaylor/slate
2. https://github.com/zenoamaro/react-quill

Both are in beta, but so is the bootstrap version we're using.

I'll try both and see which fits better.

[the remark this answers is not in the document]
Comment (giovannicimolin) [reply frontend/package.json:83]: Btw, the one you suggested also looks a bit abandoned. No commits since January.

[the remark this answers is not in the document]
Comment (giovannicimolin) [reply frontend/package.json:83]: @Agrendalath Sigh, I'm having a bunch of trouble with every editor I try to fit in here :(

[the remark this answers is not in the document]
Comment (giovannicimolin) [reply frontend/package.json:83]: Slate is too complex, and React Quill has a lot of quirks and weird bugs.

File: frontend/src/console/components/CustomizationSideMenu/CustomizationSideMenu.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (giovannicimolin) [reply frontend/src/console/components/CustomizationSideMenu/CustomizationSideMenu.tsx]: Done!

File: frontend/package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (giovannicimolin) [reply frontend/package.json]: Switched to TinyMCE, and it works wonders, has no state management bugs! 
Two downsides though: 
1. Unable to add snapshot tests (tinymce generates a random id and I didn't dig too deep on how to fix that).
2. We'll need to `eject` the project to include TinyMCE into the bundle, otherwise we'll get this pesky warning:
Image Pasted at 2020-4-9 12-00 (1)

File: frontend/src/console/components/CustomPages/CustomPages.tsx [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+                    id="show-page-switch"
+                  />
+                </Form>
+              </Col>
+            </Row>
+
+            {/* Uncomment when the backend properly passes back LMS URL */}
+            {/* <a className="pageLink" href="http://google.com">
+              http://google.com
+            </a> */}
```
Comment (Agrendalath) [inline frontend/src/console/components/CustomPages/CustomPages.tsx:157]: Nit: could we leave a placeholder link here (e.g. just a link to opencraft.com) instead of hiding it, like we did with the toggle switch?

File: frontend/src/console/components/CustomPages/CustomPages.tsx [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+            <Prompt
+              when={this.hasContentChanged()}
+              message={messages.leavePageMessage.defaultMessage}
+            />
```
Comment (Agrendalath) [inline frontend/src/console/components/CustomPages/CustomPages.tsx:162]: Nice trick :)

File: frontend/src/console/components/CustomPages/EditorConfig.tsx [hunk 6]
```diff
@@ -0,0 +1,46 @@
+/*
+ * Quill modules to attach to editor
+ * See https://quilljs.com/docs/modules/ for complete options
```
[the remark this answers is not in the document]
Comment (giovannicimolin) [reply frontend/src/console/components/CustomPages/EditorConfig.tsx:3]: No, removing.

File: frontend/packages/api_client/src/apis/V2Api.ts [hunk 7]
```diff
@@ -511,6 +511,10 @@ export class V2Api extends runtime.BaseAPI {
         let useForm = false;
         // use FormData to transmit files using content-type "multipart/form-data"
         useForm = canConsumeForm;
+        // use FormData to transmit files using content-type "multipart/form-data"
+        useForm = canConsumeForm;
+        // use FormData to transmit files using content-type "multipart/form-data"
+        useForm = canConsumeForm;
```
[the remark this answers is not in the document]
Comment (giovannicimolin) [reply frontend/packages/api_client/src/apis/V2Api.ts:517]: Unfortunately, no. I've deleted the entire folder and recreated the API client and this still shows up. This didn't happen with the previous docker image I was using.

File: frontend/packages/api_client/.openapi-generator/VERSION [hunk 8]
```diff
@@ -1 +1 @@
-4.3.1-SNAPSHOT
\ No newline at end of file
+4.2.3-SNAPSHOT
```
[the remark this answers is not in the document]
Comment (giovannicimolin) [reply frontend/packages/api_client/.openapi-generator/VERSION]: Done!

File: frontend/packages/api_client/src/runtime.ts [hunk 9]
```diff
@@ -144,7 +144,7 @@ export class Configuration {
     constructor(private configuration: ConfigurationParameters = {}) {}
 
     get basePath(): string {
-        return this.configuration.basePath != null ? this.configuration.basePath : BASE_PATH;
+        return this.configuration.basePath || BASE_PATH;
```
[the remark this answers is not in the document]
Comment (giovannicimolin) [reply frontend/packages/api_client/src/runtime.ts]: Updated my docker imaged and this was reverted.

File: frontend/package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (giovannicimolin) [reply frontend/package.json]: @Agrendalath I'll keep replying here for context.

>Does it mean that we would need to update everything for this package manually instead of using npm for this?

No we'll still use `npm` to manage packages as usual. The main change would be that we'll expose all webpack configuration and npm command scripts and store them in the repo.

This will allow us to tweak our project to better fit our needs, like enabling CSS Modules (not working because we need to add a loader to the WebPack configuration), and including additional dependencies in the bundle (like the TinyMCE imports).

For the TinyMCE change, we just need to add a few lines to `webpack.conf.js` (which is not accessible until we eject):
```javascript
      // TinyMCE
      // Must be accessible globaly for tinymce-react
      {
        test: require.resolve('tinymce/tinymce'),
        use: [ 'imports-loader?this=>window', 'exports-loader?window.tinymce' ]
      },
```

File: frontend/src/console/components/CustomPages/CustomPages.tsx (hunk 4 above)
Comment (giovannicimolin) [reply frontend/src/console/components/CustomPages/CustomPages.tsx:157]: I'm adding the LMS and studio endpoints to the API. :)

File: frontend/package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (Agrendalath) [reply frontend/package.json]: > I'll keep replying here for context.

Ok, I'm seeing the "reply" section now. Must've been some weird GH bug.

Got it, sounds reasonable then :+1:

File: frontend/src/console/components/CustomPages/CustomPages.tsx [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+    this.setState({
+      pageContent: newContent
+    });
+  };
+
+  getLMSLinkForPage = () => {
+    const instance = this.props.activeInstance;
+    const { pageName } = this.props.match.params;
+
+    if (instance && instance.data && instance.data.lmsUrl) {
```
Comment (Agrendalath) [inline frontend/src/console/components/CustomPages/CustomPages.tsx:113]: It's `undefined` for me, despite the fact that I'm having `lms_url` returned by `/api/v2/instances/openedx_config/` endpoint.

File: frontend/src/console/components/CustomPages/CustomPages.tsx [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+export class CustomPagesComponent extends React.PureComponent<Props, State> {
+  constructor(props: Props) {
+    super(props);
+
+    this.state = {
+      pageContent: this.getPageContentFromState()
+    };
+  }
+
+  componentDidUpdate = (prevProps: Props) => {
```
Comment (Agrendalath) [inline frontend/src/console/components/CustomPages/CustomPages.tsx:46]: Shouldn't we define it like `componentDidUpdate(prevProps: Props) {` to overload the parent's class method?

Comment (giovannicimolin) [reply frontend/src/console/components/CustomPages/CustomPages.tsx:46]: It binds in the same way afaik, so I went with the `= () =>` to keep all implementations consistent inside this component.

Comment (giovannicimolin) [reply frontend/src/console/components/CustomPages/CustomPages.tsx:46]: Do you want me to change it to `componentDidUpdate(prevProps: Props) {`?

File: frontend/src/console/components/CustomPages/CustomPages.tsx (hunk 10 above)
Comment (giovannicimolin) [reply frontend/src/console/components/CustomPages/CustomPages.tsx:113]: @Agrendalath You have to rebuild the API client, otherwise the `lms_url` field won't be translated to `lmsUrl`.

File: frontend/src/console/components/CustomPages/CustomPages.tsx (hunk 11 above)
Comment (Agrendalath) [reply frontend/src/console/components/CustomPages/CustomPages.tsx:46]: Ah, I see that my IDE mistakenly took the TypeScript version from `api_client` `v2.9.2` instead of the base one `v3.6.4`, which resulted in [this error](https://github.com/microsoft/TypeScript/issues/17113). It's fine to keep it this way. Sorry for the confusion!

File: frontend/src/global/constants.ts [hunk 12]
```diff
@@ -56,3 +63,21 @@ export const REGISTRATION_STEPS = [
...
+  'contact',
+  'donate',
+  'tos',
+  'honor',
+  'privacy'
+];
+
+export const LMS_CUSTOM_PAGE_LINK_MAP: StringIndexedArray = {
+  about: 'about',
+  contact: 'suport/contact_us',
```
Comment (Agrendalath) [inline frontend/src/global/constants.ts:78]: Nit: typo in `support`.

File: frontend/src/console/components/CustomPages/CustomPages.tsx (hunk 10 above)
Comment (Agrendalath) [reply frontend/src/console/components/CustomPages/CustomPages.tsx:113]: Ah, forgot about this step. Works great now!

File: frontend/src/global/constants.ts (hunk 12 above)
Comment (giovannicimolin) [reply frontend/src/global/constants.ts]: Nice catch!
